### PR TITLE
Refactor imagecreatefrombmp

### LINKED ISSUE #26758: Bug with 1 bit bmp
### Steps to reproduce
1. Download http://cloud.acomirei.ru/index.php/s/0sHvBwXpMKxYxTK and unzip
2. Upload 3.bmp to the server using web ui

### Expected behaviour
The server generates correct preview image.

### Actual behaviour
Tons of

> {"reqId":"+iIwdobvVxHII3QewZ5J","remoteAddr":"10.0.0.101","app":"PHP","message":"unpack(): Type n: not enough input, need 2, have 1 at \/var\/www\/html\/owncloud\/lib\/private\/legacy\/image.php#748","level":3,"time":"2016-12-02T11:04:27+00:00","method":"GET","url":"\/index.php\/apps\/files\/api\/v1\/thumbnail\/256\/256\/2_\u0420\u0430\u0441\u0447\u0435\u0442 \u0447\u0435\u0440\u0432\u044f\u0447\u043d\u043e\u0439 \u043f\u0435\u0440\u0435\u0434\u0430\u0447\u0438\/3.bmp","user":"denis"}
> {"reqId":"+iIwdobvVxHII3QewZ5J","remoteAddr":"10.0.0.101","app":"PHP","message":"unpack(): Type n: not enough input, need 2, have 1 at \/var\/www\/html\/owncloud\/lib\/private\/legacy\/image.php#748","level":3,"time":"2016-12-02T11:04:27+00:00","method":"GET","url":"\/index.php\/apps\/files\/api\/v1\/thumbnail\/256\/256\/2_\u0420\u0430\u0441\u0447\u0435\u0442 \u0447\u0435\u0440\u0432\u044f\u0447\u043d\u043e\u0439 \u043f\u0435\u0440\u0435\u0434\u0430\u0447\u0438\/3.bmp","user":"denis"}

The server hangs and doesn't stop to populate a log file - when i found the bug the size of the log file was about 50GB.

### Server configuration
**Operating system**: 
Centos 7.0.

**Web server:**
Cherokee.

**Database:**
MySQL.

**PHP version:**
5.6.

**ownCloud version:** (see ownCloud admin page)
9.1.2.

**Updated from an older ownCloud or fresh install:**
Updated from an older.

**Where did you install ownCloud from:**
Using owncloud repo.

**Signing status (ownCloud 9.0 and above):**

```
No errors have been found.
```

**List of activated apps:**

```
Enabled:
  - activity: 2.3.2
  - comments: 0.3.0
  - dav: 0.2.7
  - federatedfilesharing: 0.3.0
  - federation: 0.1.0
  - files: 1.5.1
  - files_pdfviewer: 0.8.1
  - files_sharing: 0.10.0
  - files_texteditor: 2.1
  - files_trashbin: 0.9.0
  - files_versions: 1.3.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - notifications: 0.3.0
  - provisioning_api: 0.5.0
  - systemtags: 0.3.0
Disabled:
  - encryption
  - external
  - files_antivirus
  - files_external
  - gallery
  - templateeditor
  - updatenotification
  - user_external
  - user_ldap
```

**The content of config/config.php:**

```
{
    "system": {
        "instanceid": "ocikxtf5b3ao",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.acomirei.ru"
        ],
        "datadirectory": "\/var\/www\/html\/owncloud\/data",
        "overwrite.cli.url": "http:\/\/cloud.acomirei.ru",
        "dbtype": "mysql",
        "version": "9.1.2.5",
        "dbname": "c_user",
        "dbhost": "10.0.0.101",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "mail_smtpmode": "smtp",
        "mail_domain": "cloud.acomirei.ru",
        "mail_from_address": "admin",
        "mail_smtphost": "10.0.0.101",
        "theme": "",
        "loglevel": 3,
        "maintenance": false,
        "preview_libreoffice_path": "\/usr\/bin\/libreoffice",
        "updatechecker": false
    }
}
```

**Are you using external storage, if yes which one:**
No.

**Are you using encryption:**
No.

**Are you using an external user-backend, if yes which one:**
No.

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
Handle reading non-existing characters for any bit depth while creating an image from bmp

## Related Issue
Fixes https://github.com/owncloud/core/issues/26758

## Motivation and Context
Solves flooding the log file with
```
PHP Warning:  unpack(): Type n: not enough input, need 2, have 1 
```
1 entry per each missing pixel

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by VicDeo** [comment]
It's better to view changes this way:
https://github.com/owncloud/core/pull/26849/files?w=1

**Comment by PVince81** [comment]
Please add unit tests

**Comment by VicDeo** [comment]
@PVince81 for a private method? :)

**Comment by PVince81** [comment]
Yes, you even increased the complexity of that method.
You can use `TestCase::invokePrivate`.

Or if that's not acceptable, then write an integration test with matching image files... I guess it's much easier to write unit tests here.

**Comment by VicDeo** [comment]
@PVince81  this method a hell lot of things and can be extracted into a separate class.
In the past life it was a function:
```
 	/**
 	 * Create a new image from file or URL
 	 *
 	 * @link http://www.programmierer-forum.de/function-imagecreatefrombmp-laeuft-mit-allen-bitraten-t143137.htm
 	 * @version 1.00
 	 * @param string $fileName <p>
 	 * Path to the BMP image.
 	 * </p>
 	 * @return bool|resource an image resource identifier on success, <b>FALSE</b> on errors.
 	 */
 	private function imagecreatefrombmp($fileName) {
```

**Comment by PVince81** [comment]
It's a good opportunity to add missing tests.

If you feel like the method does too much then split it into smaller functions and add tests for those.

Most important point is that we don't add tests where they are missing we'll likely have more regressions in the future. I've seen enough cases of missing/neglected/incomplete unit tests that lead to criticial regressions.

**Comment by butonic** [comment]
Blame it on me. Thx for digging into it & trying to understand what it does. 🙇

**Comment by VicDeo** [comment]
@butonic np, I'm refactoring this method into class now. :)

**Comment by PVince81** [comment]
Code looks fine. @Temtaime can you help testing ?

**Comment by Temtaime** [comment]
I'll look into this.

File: lib/private/Image/BmpToResource.php [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+		}
+		$rawDibHeader = $this->readFile($dibHeaderSizeArray['headersize'] - 4);
+		$dibHeader = @unpack('Vwidth/Vheight/vplanes/vbits/Vcompression/Vimagesize/Vxres/Vyres/Vcolors/Vimportant', $rawDibHeader);
+
+		// fixup colors
+		$dibHeader['colors'] = $dibHeader['colors'] === 0 ? pow(2, $dibHeader['bits']) : $dibHeader['colors'];
+
+		// fixup imagesize - it can be zero
+		if ($dibHeader['imagesize'] < 1) {
+			$dibHeader['imagesize'] = $this->fixImageSize($header);
```
Comment (DeepDiver1975) [inline lib/private/Image/BmpToResource.php]: $header is undefined

File: tests/lib/Image/BmpToResourceTest.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace Test\Image;
+
+use OC\Image\BmpToResource;
+use Test\TestCase;
+
+class BmpToResourceTest extends TestCase {
+	const CLASS_NAME = 'OC\Image\BmpToResource';
```
Comment (DeepDiver1975) [inline tests/lib/Image/BmpToResourceTest.php]: BmpToResource::class

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by VicDeo** [comment]
@DeepDiver1975 thanks for reviewing. Updated.

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by DeepDiver1975** [comment]
Code coverage is at 75% - can we get some more? THX

**Comment by VicDeo** [comment]
@DeepDiver1975 ok, let it be 88%

**Pull request merged**